File: JavaBackend/src/main/java/repositories/UsersRepoImpl.java
package repositories;

import models.Users;
import util.JDBCConnection;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class UsersRepoImpl implements UsersRepo {

    public static Connection conn = JDBCConnection.getConnection();

    public UsersRepoImpl(){ }

    @Override
    public Users getUser(int id) {
        String sql = "SELECT * FROM users WHERE userid = ?";

        try {
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setInt(1, id);

            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                return buildUser(rs);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public List<Users> getAllUsers() {
        String sql = "SELECT * FROM users";

        try {
            PreparedStatement ps = conn.prepareStatement(sql);

            ResultSet rs = ps.executeQuery();

            List<Users> users = new ArrayList<>();

            while(rs.next()) {
                users.add(buildUser(rs));
            }

            return users;

        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public Users addUser(Users u) {
        String sql = "INSERT INTO users VALUES (DEFAULT,?,?,?,?) RETURNING *";

        try {
            PreparedStatement ps = conn.prepareStatement(sql);

            ps.setString(1, u.getFirstname());
            ps.setString(2, u.getLastname());
            ps.setString(3, u.getEmail());
            ps.setString(4, u.getPhone());

            ResultSet rs = ps.executeQuery();

            if(rs.next()) {
                return buildUser(rs);
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public Users updateUser(Users change) {
        try {
            String sql = "UPDATE users SET firstname=?, lastname=?, email=?, phone=? WHERE userid = ? RETURNING *";
            PreparedStatement ps = conn.prepareStatement(sql);

            ps.setString(1, change.getFirstname());
            ps.setString(2, change.getLastname());
            ps.setString(3, change.getEmail());
            ps.setString(4, change.getPhone());
            ps.setInt(5, change.getUserID());

            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                return buildUser(rs);
            }


        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public Users deleteUser(int id) {
        try {
            String sql = "DELETE FROM users WHERE userid = ? RETURNING *";
            PreparedStatement ps = conn.prepareStatement(sql);

            ps.setInt(1, id);
            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                return buildUser(rs);
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    private Users buildUser(ResultSet rs) throws SQLException {
        Users u = new Users();
        u.setUserID(rs.getInt("userid"));
        u.setFirstname(rs.getString("firstname"));
        u.setLastname(rs.getString("lastname"));
        u.setEmail(rs.getString("email"));
        u.setPhone(rs.getString("phone"));

        return u;
    }

}

File: ProjectTwo(Python)/repositories/users_repo_impl.py
from models.users import Users
from repositories.user_repo import UserRepo
from exceptions.resource_not_found import ResourceNotFound
from util.db_connection import connection


def _build_user(record):
    return Users(user_id=record[0], first_name=record[1], last_name=record[2], email=record[3], phone=record[4])


class UserRepoImpl(UserRepo):

    def create_user(self, user):
        sql = "INSERT INTO Users VALUES (DEFAULT, %s, %s, %s, %s) RETURNING *"
        cursor = connection.cursor()
        cursor.execute(sql, [user.user_id, user.first_name, user.last_name, user.email, user.phone])
        connection.commit()
        record = cursor.fetchone()
        return _build_user(record)

    def get_user(self, user_id):
        sql = "SELECT * FROM Users WHERE userID = %s"
        cursor = connection.cursor()
        cursor.execute(sql, [user_id])
        record = cursor.fetchone()

        if record:
            return _build_user(record)
        else:
            raise ResourceNotFound(f"User with ID: {user_id} - Not Found")

    def get_all_users(self):
        sql = "SELECT * FROM Users"
        cursor = connection.cursor()
        cursor.execute(sql)
        records = cursor.fetchall()
        user_list = [_build_user(record) for record in records]
        return user_list

    def update_user(self, change):
        sql = "UPDATE Users SET firstname = %s, lastname = %s, email = %s, phone = %s RETURNING *"
        cursor = connection.cursor()
        cursor.execute(sql, [change.first_name, change.last_name, change.email, change.phone])
        connection.commit()
        record = cursor.fetchone()
        return _build_user(record)

    def del_user(self, user_id):
        sql = "DELETE FROM Users WHERE userID = %s"
        cursor = connection.cursor()
        cursor.execute(sql, [user_id])
        connection.commit()
        return f"User ID: {user_id} - Deleted"


def _test():
    ur = UserRepoImpl()

    print(ur.get_user(1))


if __name__ == '__main__':
    _test()
